### DISCUSSION TITLE
Pydantic 2.6.4 -> 2.7.0: expected enum but got str serialized value may not be as expected

### DISCUSSION THREAD

**Discussion by hhartzer** [pydantic/pydantic#9270] [Pydantic V2]
I'd like to open this as an issue, but I don't have a reproducible example yet. Hoping someone else may have seen this.

FastAPI 0.101.1, sqlmodel 0.0.16, Python 3.11.

Going from Pydantic 2.6.4 to 2.7.0, some heavy endpoints, maybe somewhat randomly (and more often than not), will spew `expected enum but got str serialized value may not be as expected` over and over. It seems to be able to lock thing up for minutes at a time, if left alone.

I do have some enums. I suspect this is related to these:

#9045

#9064

Which showed up in 2.7.0.

When I reverted (solely) back to 2.6.4, all was well.

This might definitely be my bug in some way, but I feel like this is a breaking change to some degree or another.

**Comment by inyutin** [comment]
Interestingly, we have very same issue on pydantic 2.6.4 :thinking: 
Python 3.11, fastapi 0.110.1

**Reply by hhartzer** [reply]
Did your pydantic-core version change with it?

**Reply by tamird** [reply]
Indeed we are on pydantic-core 2.18.1.

**Reply by hhartzer** [reply]
I think that pydantic 2.7.0 is supposed to go with pydantic-core 2.18.1, while pydantic 2.6.4 should go with 2.16.3.

Can you downgrade pydantic-core and see if it fixes this?

**Reply by tamird** [reply]
I found the problem and fixed it. We were indeed doing the wrong thing.

**Comment by junoriosity** [comment]
@hhartzer I also encountered this issue, is there a way to overcome that impediment?

**Reply by hhartzer** [reply]
You can downgrade. I think the best thing would be to make a small, reproducible example so we can make a bug report with it. I haven't done that yet.

**Comment by miketheman** [comment]
Here's an example that's reproducible, but I suspect it's an operator error.

```python
from enum import Enum

from pydantic import BaseModel


class FruitEnum(str, Enum):
    pear = 'pear'
    banana = 'banana'


class CookingModel(BaseModel):
    fruit: FruitEnum | None = None
  
# no warnings emitted since the enum is used in the model construction
print(CookingModel(fruit='banana').model_dump_json(indent=2))

cm = CookingModel()
cm.fruit = 'pear'
print(cm.model_dump_json(indent=2))
# warning: "Expected `enum` but got `str` - serialized value may not be as expected"
```

After starting a bug report, I realized that it's partially my fault, since I'm assigning the value of `fruit` outside the construction step, which probably skips validation.

Changing to:

```python
cm.fruit = FruitEnum('pear')
```

resolves the warning, as it's using the string `'pear'` to look up the enum value from `FruitEnum` and returns the correct type for the dump renderer.

**Reply by hhartzer** [reply]
Awesome, thank you!

I do feel like if it's a `str, Enum`, a matching `str` should be counted as well.

**Reply by junoriosity** [reply]
@hhartzer Is there some progress on that matter?

**Reply by hhartzer** [reply]
@junoriosity I haven't looked into this in a while. I've just been using the older version for now.

**Comment by uxduck** [comment]
This error was fixed for me by using using `model_config = ConfigDict(use_enum_values=True)`.

I get messages (class instances, with fields that are integers) that i don't need to validate (it has it's own validation) but I want to convert it into my own pydantic models that use enums. I don't want to JSON serialize and do `model_validate_json`, so I'm using model_construct.

```python
# Imagine I get a class instance which I converted into a dictionary.
external_message = {"cats": 0}

# These are some classes I define. I know the external messages respect this API.
class Cat(enum.IntEnum):
    lemon = 0
    winston = 1

class ExampleMain(BaseModel):
    model_config = ConfigDict(use_enum_values=True) # <-- 👈🏼️ this was the fix

    cats: Cat
    id: int

# I use model_construct because I don't want to validate my data I trust, but my integers are e.g. `1`, not enums `Cat.lemon`
e1 = ExampleMain.model_construct(**external_message, id=1)
e1.model_dump_json()
```

`model_dump_json()` above emits the following warning if we don't use the model_config I used above:
```
# WARNING: Expected `enum` but got `int` with value `0` - serialized value may not be as expected
```

**Reply by hhartzer** [reply]
This worked for me. Thank you!
